streaming/base/dataset.py

```diff
@@ -367,201 +456,166 @@ def _get_progress(self, world: World) -> Tuple[int, int]:
+        return np.where(big_ids != -1, small_per_big[big_ids], -1)
+
+    def _get_worker_samples(self, world: World, epoch: int,
+                            sample_in_epoch: int) -> NDArray[np.int64]:
+        """Get this worker's partition of this epoch's sample space.
 
         Args:
-            zip_info (FileInfo): Compressed file info.
-            zip_filename (str): Compressed filename.
-            raw_filename (str): Decompressed filename.
-            compression (str, optional): Compression algorithm.
-        """
-        # Load compressed.
-        data = open(zip_filename, 'rb').read()
-
-        # Validate what was downloaded.
-        if self.validate_hash:
-            if get_hash(self.validate_hash, data) != zip_info.hashes[self.validate_hash]:
-                raise ValueError(f'Checksum failure: {zip_filename}')
-
-        # Decompress and save that.
-        data = decompress(compression, data)  # pyright: ignore
-        tmp_filename = raw_filename + '.tmp'
-        with open(tmp_filename, 'wb') as out:
-            out.write(data)
-        os.rename(tmp_filename, raw_filename)
-
-        # Maybe remove compressed to save space.
-        if not self.keep_zip:
-            os.remove(zip_filename)
-
-    def _download_shard_part(self,
-                             raw_info: FileInfo,
-                             zip_info: Optional[FileInfo] = None,
-                             compression: Optional[str] = None) -> None:
-        """Download shard data given metadata for the raw and compressed versions of it.
-
-        MDS format uses joint shards (ie, one file per shard). Other formats supported by streaming
-        use split shards (ie, shard data lives in two files per shard: the raw data itself and
-        metadata in a separate file).
+            world (World): World state.
+            epoch (int): Which epoch it is.
+            sample_in_epoch (int): Where we are in the epoch.
 
-        Args:
-            raw_info (FileInfo): Raw file info.
-            zip_info (FileInfo, optional): Zip file info. Defaults to ``None``.
-            compression (str, optional): Compression algorithm used for zip_info. Defaults to
-                ``None``.
+        Returns:
+            Optional[NDArray[np.int64]]: Our partition of the epoch.
         """
-        # If the local raw file already exists, this is a no-op.
-        raw_filename = os.path.join(self.local, self.split, raw_info.basename)
-        if os.path.isfile(raw_filename):
-            return
+        # Do expensive work that may use all the cores just once, in the local leader.
+        if world.is_local_leader:
+            # Generate the global arrangement of samples for this epoch.
+            sample_ids = self._generate_epoch(world, epoch, sample_in_epoch)
+
+            # Save the generated epoch shape to shared memory.
+            name = f'{self._prefix}_epoch_shape'
+            size = 5 * np.int64().nbytes
```

Comment:
Good to document this magic `5` number !!

Reply:
refactored